Set up `zppy` executable

forsyth2 opened the pull request:
Set up `zppy` executable. Resolves #30.

forsyth2 commented on conda/dev.yml:
```diff
@@ -0,0 +1,9 @@
+name: zppy_dev
```
@tomvothecoder Do `dev.yml` and `meta.yaml` look good to you? I modeled them off their respective files for `zstash` (https://github.com/E3SM-Project/zstash/tree/master/conda). Note that we haven't addressed #10 (CI/CD workflow) yet, so I took out the developer tools.

tomvothecoder replied:
@forsyth2 The template looks good, just need to add dependencies for both files as mentioned in the other comments.

forsyth2 commented on conda/dev.yml:
```diff
@@ -0,0 +1,9 @@
+channels:
+  - conda-forge
+  - defaults
+dependencies:
```
@golaz I'm not sure what specific dependencies `zppy` has. I know it doesn't have many since the generated bash scripts do most of the actual work.

forsyth2 reviewed (commented):
@xylar I went through your bulleted list on #30. Thanks! 

Am I on the right track with this pull request? I imagine I'll need to tweak at least a few things.

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,32 @@
+  number: 0
+  script: "{{ PYTHON }} -m pip install . --no-deps -vv"
+  noarch: python
+
```
```suggestion
  entry_points:
    - zppy = zppy.__main__:main
    
```

xylar commented on setup.py:
```diff
@@ -0,0 +1,11 @@
+    author_email="forsyth2@llnl.gov, golaz1@llnl.gov",
+    description="Post-processing software for E3SM",
+    packages=find_packages(exclude=["*.test", "*.test.*", "test.*", "test"]),
+    entry_points={"console_scripts": ["zppy=zppy.main:main"]},
```
```suggestion
    entry_points={"console_scripts": ["zppy=zppy.__main__:main"]},
```

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,32 @@
+    - pip
+
+  run:
+    - python >=3.6
```
I think `zppy` has a bunch more dependencies that should explicitly be added.  Here's what I see so far (please keep alphabetized:
* `cdms2`
* `cdutil`
* `cdtime`
* `jinja2`
* `matplotlib-base`
* `netcdf4`
* `numpy`

I have a feeling this is an incomplete list building, installing and running the package on its own may catch others.

xylar replied:
As many as possible of these should also be added to the `install_requires` argument to `setup()` in `setup.py`.

xylar replied:
You just need to add:
```
    - configobj
    - jinja2
```

xylar reviewed (commented):
@forsyth2, there are going to be a ton of other issues still to deal with.  Here are a few that come to mind:
* Every directory in the package needs an `__init__.py`, even if it's just empty.
* All non-python files will need to be explicitly included in the package files (see the [compass package](https://github.com/MPAS-Dev/compass/blob/master/setup.py#L48-L50) for an example)
* Imports will either need to be made relative (`from .mpas_analysis import mpas_analysis`) or absolut (`from zppy.mpas_analysis import mpas_analysis`).  I recommend absolute.
* References to non-python files will have to be referenced differently, preferably with `importlib.resources`, see [this example](https://github.com/MPAS-Dev/compass/blob/864ebf7425d5f076d4c18673b43d102308da7071/compass/landice/tests/enthalpy_benchmark/B/__init__.py#L44-L46) from compass

It would be nice to have a video chat about this at some point involving whoever is interested.  I'm at EGU this week but maybe next week?

forsyth2 commented:
@xylar Thanks for the comments. Yes, we can figure out a time next week to discuss further.

forsyth2 commented on zppy/__main__.py:
```diff
@@ -0,0 +1,82 @@
+from amwg import amwg
+from mpas_analysis import mpas_analysis
+
+def main():
+  
+  # Command line parser
+  parser = argparse.ArgumentParser(description='Launch E3SM post-processing tasks', usage="zppy -c <config>")
+  parser.add_argument(
+    '-c', '--config', type=str, help='configuration file',required=True)
+  args = parser.parse_args()
```
We do plan to add a linter (probably as part of #10) -- but I doubt that will be completed before the first release.

xylar replied:
That's obviously up to you and @golaz.  In my experience, it's much less confusing to go through the git history if reformatting happens as early in the overall process as possible.

forsyth2 commented on zppy/mpas_analysis.py:
```diff
@@ -1,14 +1,17 @@
     # Initialize jinja2 template engine
     templateLoader = jinja2.FileSystemLoader(searchpath=config['default']['templateDir'])
     templateEnv = jinja2.Environment( loader=templateLoader )
+    # TODO: need to get path from package passed to it (importlib.resource.path)
+    #with resources.path('zppy.templates', 'mpas_analysis.bash') as target:
+    #    os.symlink(str(target), '{}/'.format('???'))
```
@xylar I was trying to emulate https://github.com/MPAS-Dev/compass/blob/864ebf7425d5f076d4c18673b43d102308da7071/compass/landice/tests/enthalpy_benchmark/B/__init__.py#L44-L46, but it seems like `zppy -c <config-file>` works fine without `importlib.resource.path`. Is that just a quirk of how I'm testing?

xylar replied:
I think this is definitely a requirement.  I do guess that things are working for you because of how you are testing. Could you send me (e.g. on Slack) the procedure your're using for testing, including your config file so I can also do some testing?

xylar replied:
Hmm, I see now why it's working.  I still think using `importlib.resource.path` is the preferred way but I'll comment below.

forsyth2 reviewed (commented):
@xylar The `zppy` executable now works for me locally, but I have a few questions (see comments). Also, I think I might break off the `conda build` component into its own pull request.

xylar commented on conda/dev.yml:
```diff
@@ -0,0 +1,11 @@
+  - defaults
+dependencies:
+  # Base
+  - configobj=5.0.6
+  - jinja2=2.11.3
+  - pip=21.0.1
+  - python=3.7.10
```
Okay, that sounds good.

xylar commented on zppy/__main__.py:
```diff
@@ -0,0 +1,88 @@
+      pass
+
+  # Determine machine to decide which header files to use
+  tmp = os.getenv('HOSTNAME')
+  if tmp.startswith('compy'):
+      machine = "compy"
+      environment_commands = 'source /share/apps/E3SM/conda_envs/load_{}_e3sm_unified.sh'.format(config['default']['e3sm_unified'])
+  elif tmp.startswith('cori'):
+      machine = "cori"
+      environment_commands = 'source /global/cfs/cdirs/e3sm/software/anaconda_envs/load_{}_e3sm_unified.sh'.format(config['default']['e3sm_unified'])
+  elif tmp.startswith('blues'):
+      machine = "anvil"
+      environment_commands = 'source /lcrc/soft/climate/e3sm-unified/load_{}_e3sm_unified.sh'.format(config['default']['e3sm_unified'])
+  elif tmp.startswith('chr'):
+      machine = "chrysalis"
+      environment_commands = 'source /lcrc/soft/climate/e3sm-unified/load_{}_e3sm_unified.sh'.format(config['default']['e3sm_unified'])
+  config['default']['machine'] = machine
+  if 'environment_commands' not in config['default'].keys():
+    config['default']['environment_commands'] = environment_commands
```
@forsyth2, this is fine.  I think your point above makes clear to me that we do not necessarily want the scripts to activate the same environment during testing that `zppy` is being run from.  You do not want the full e3sm-unified installed in your test enviornment.

xylar commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+
+data_files = package_files('zppy', prefixes=[], extensions=['bash', 'csh', 'ini'])
+
+setup(
+    name="zppy",
+    version="1.0.0",
+    author="Ryan Forsyth, Chris Golaz",
+    author_email="forsyth2@llnl.gov, golaz1@llnl.gov",
+    description="Post-processing software for E3SM",
+    packages=find_packages(exclude=["*.test", "*.test.*", "test.*", "test"]),
+    package_data={'': data_files},
+    entry_points={"console_scripts": ["zppy=zppy.__main__:main"]},
+)
```
```suggestion
setup(
    name="zppy",
    version="1.0.0",
    author="Ryan Forsyth, Chris Golaz",
    author_email="forsyth2@llnl.gov, golaz1@llnl.gov",
    description="Post-processing software for E3SM",
    python_requires='>=3.7',
    intall_requires=['configobj', 'jinja2'],
    packages=find_packages(exclude=["*.test", "*.test.*", "test.*", "test"]),
    package_data={'': data_files},
    entry_points={"console_scripts": ["zppy=zppy.__main__:main"]},
)
```
I'm guessing on python version constraints.  If there are version constraints on the other 2 dependencies, please add them

xylar commented:
I'm trying to test this using a variant of @golaz's config file here:
https://e3sm-project.github.io/zppy/docs/html/getting_started.html
but I'm running into error after error:
```
  File "/gpfs/fs1/home/ac.xylar/e3sm_work/zppy/executable/zppy/templates/e3sm_diags.bash", line 62, in top-level template code
    ts_dir={{ output }}/post/atm/{{ grid }}/ts/monthly/{{ '%dyr' % (ts_num_years) }}
jinja2.exceptions.UndefinedError: 'ts_num_years' is undefined
```
then, when I add that:
```
  File "/gpfs/fs1/home/ac.xylar/e3sm_work/zppy/executable/zppy/e3sm_diags.py", line 53, in e3sm_diags
    c['climo_diurnal_subsection'], c['year1'],c['year2'])))
KeyError: 'climo_diurnal_subsection'
```
@forsyth2 or @golaz, could you give me an equivalent to those instructions that should work for this branch?

xylar commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+                paths.append(os.path.join('..', path, filename))
+    return paths
+
+data_files = package_files('zppy', prefixes=[], extensions=['bash', 'csh', 'ini'])
```
```suggestion
data_files = package_files('zppy', prefixes=[], extensions=['bash', 'csh', 'ini', 'sh'])
```

xylar commented on zppy/__main__.py:
```diff
@@ -0,0 +1,88 @@
+  args = parser.parse_args()
+
+  # Subdirectory where templates are located
+  templateDir = os.path.join( os.path.dirname(__file__), 'templates' )
```
While this is not an ideal practice, this will work for finding `templates` in circumstances we work in.  See https://pyoxidizer.readthedocs.io/en/v0.5.0/packaging_pitfalls.html#reliance-on-file for some pitfalls and why `importlib.resources` would be preferable in general.

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/E3SM-Project/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: ab6dee537f8252a9f244f01b725046dd64748d68
```
For testing, this can be:
```
  path: ..
```

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  run:
+    - python >=3.6
+
+test:
+  imports:
+    - zppy
+  commands:
+    - zppy --help
```
```suggestion
test:
  requires:
    - pip
  imports:
    - zppy
  commands:
    - zppy --help
    - pip check
```
`pip check` is a nice thing to add.

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  commands:
+    - zppy --help
+
+about:
+  home: https://github.com/E3SM-Project/zppy
```
This needs to get filled out: 
https://github.com/conda-forge/e3sm_diags-feedstock/blob/master/recipe/meta.yaml#L51-L66

xylar commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+    version="1.0.0",
+    author="Ryan Forsyth, Chris Golaz",
+    author_email="forsyth2@llnl.gov, golaz1@llnl.gov",
+    description="Post-processing software for E3SM",
+    packages=find_packages(exclude=["*.test", "*.test.*", "test.*", "test"]),
```
```suggestion
    description="Post-processing software for E3SM",
    python_requires='>=3.6',
    intall_requires=['configobj', 'jinja2'],
    packages=find_packages(exclude=["*.test", "*.test.*", "test.*", "test"]),
```
I'm guessing on python version constraints.  If there are version constraints on the other 2 dependencies, please add them

xylar commented:
@forsyth2, in case you at least want to test the conda build on this branch (which I highly recommend), once you've made the changes I recommended to `meta.yaml` and you've temporarily replaced `url`, etc. with `path: ..`, to build and test, just do:
```
source ~/miniconda3/etc/profile.d/conda.sh
conda activate
conda build conda
conda create -y -n zppy_package --use-local zppy
conda activate zppy_package
zppy -c <config>
```
This will help to find files that are missing from the package (not in `data_files` in `setup.py`) and will simulate how things will be in E3SM-Unified most closely.

forsyth2 commented:
@ Xylar I addressed some of your comments in a new commit.

- I think I will deal with the `importlib.resources` issue in a separate pull request since the code still works and I want to make sure we get this pull request (and other features) in to v1.0.0.

- I'm following your steps

```
source ~/miniconda3/etc/profile.d/conda.sh
conda activate
conda build conda
conda create -y -n zppy_package --use-local zppy
conda activate zppy_package
zppy -c <config>
```
but I get `ModuleNotFoundError: No module named 'configobj'` on the `zppy -c` line. I don't understand why that would be the case since the `dev` environment works and I have the dependencies in the `meta` file too.

- The docs aren't up to date. You can use the following configuration for a simple run:

```
[default]
input = /lcrc/group/e3sm/ac.forsyth2/E3SM_simulations/20210122.v2_test01.piControl.ne30pg2_EC30to60E2r2-1900_ICG.chrysalis
input_subdir = archive/atm/hist
output = /lcrc/group/e3sm/ac.forsyth2/E3SM_simulations/20210122.v2_test01.piControl.ne30pg2_EC30to60E2r2-1900_ICG.chrysalis
case = 20210122.v2_test01.piControl.ne30pg2_EC30to60E2r2-1900_ICG.chrysalis
www = /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/E3SM/v2/beta
e3sm_unified = latest
partition = compute
ref_start_yr = 1979
ref_final_yr = 2016

[climo]
active = True
years = "1:20:20",
# Use default vars

  [[ atm_monthly_180x360_aave ]]
  mapping_file = /home/ac.zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  frequency = "monthly"

  [[ atm_monthly_diurnal_8xdaily_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h4"
  mapping_file = /home/ac.zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  vars = "PRECT"
  frequency = "monthly_diurnal_8xdaily"

[ts]
active = True
years = "1:20:10",

  [[ atm_monthly_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h0"
  frequency = "monthly"
  mapping_file = /home/ac.zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  # Use default vars

[glb]
active = False

[e3sm_diags]
active = True
years = "1:20:20",
ts_num_years = 10
environment_commands="source /home/ac.forsyth2/miniconda3/etc/profile.d/conda.sh; conda activate e3sm_diags_env_dev"

  [[ atm_monthly_180x360_aave ]]
  short_name = '20210122.v2_test01.piControl.ne30pg2_EC30to60E2r2-1900_ICG.chrysalis'
  grid = '180x360_aave'
  reference_data_path = '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/climatology'
  obs_ts = '/lcrc/soft/climate/e3sm_diags_data/obs_for_e3sm_diags/time-series'
  dc_obs_climo = '/lcrc/group/e3sm/public_html/e3sm_diags_test_data/unit_test_complete_run/obs/climatology'
  climo_diurnal_subsection = "atm_monthly_diurnal_8xdaily_180x360_aave"

[e3sm_diags_vs_model]
active = False

[amwg]
active = False

[mpas_analysis]
active = False
walltime = "24:00:00"
parallelTaskCount = 6
ts_years = "1-50", "1-100",
enso_years = "11-50", "11-100",
climo_years ="21-50", "51-100",
mesh = "EC30to60E2r2"

[global_time_series]
active = False
years = "1-100",
ts_num_years = 10
figstr=coupled_v2_test01
moc_file=mocTimeSeries_0001-0100.nc
experiment_name=20210122.v2_test01.piControl.chrysalis
ts_years = "1-50", "1-100",
```

xylar commented:
> - I'm following your steps...but I get `ModuleNotFoundError: No module named 'configobj'` on the `zppy -c` line. I don't understand why that would be the case since the `dev` environment works and I have the dependencies in the `meta` file too.

This is almost certainly because you haven't run:
```
conda config --add channels conda-forge
conda config --set channel_priority strict
```
See https://conda-forge.org/#about.  This should become an automatic part of installing miniconda3.

xylar commented:
> The docs aren't up to date. Here is my configuration file for a simple, fast-completing run

Thanks!

xylar commented on conda/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+    - pip
+  run:
+    - python >=3.6
+    - configobj=5.0.6
+    - jinja2=2.11.3
```
@forsyth2, while it is okay (but not my preferred practice) to pin these in the development environment, we will have major issues if we pin them this precisely in the conda recipe.  There should only be constraints on these packages if there are known incompatibilities with past versions or expected incompatibilities with future versions.  This depends on knowing the packages themselves and their histories.

I happen to know that `jinja2` 3.0.0 (why not `jinja3`, I can't say) was released a few days ago.  I would therefore expect that it might be important to test that `jinja2` 3.0.0 works well or do `jinja2 <3.0.0`.

conda-forge is picky about linting recipes.  If you want to pin a specific version, you either need no `=` or a ` ==` (note the space).

```suggestion
    - configobj 
    - jinja2
```
or
```suggestion
    - configobj >=5.0.0,<6.0.0
    - jinja2 >=2.0.0,<3.0.0
```

xylar replied:
I looked into `configobj` and it hasn't been updated since 2014, which is alarming.  But that does suggest that pinning to the latest version will probably do no harm.

xylar commented:
> I think I will deal with the `importlib.resources` issue in a separate pull request since the code still works and I want to make sure we get this pull request (and other features) in to v1.0.0.

That seems totally reasonable to me, now that I see how things are working now.  The current approach will work fine for us.  I mostly want to make you (and @golaz) aware that this is not the preferred approach for finding files in a python package for future reference.  If this gets followed up on in the future, that would be great!

xylar reviewed (approved):
Thanks for the config file.  With that (modified to point to my scratch space, of course), it seems to have worked great.

I built the conda package and used that for my testing.  (I dropped the pins on `configobj` and `jinja2` in my testing, as I suggested above.)

xylar commented:
I spoke a bit too soon.  It seems my file `e3sm_diags_atm_monthly_180x360_aave_model_vs_obs_0001-0020.bash` is missing the line for activating E3SM-Unified.  I commented out:
```
#environment_commands="source /home/ac.forsyth2/miniconda3/etc/profile.d/conda.sh; conda activate e3sm_diags_env_dev"
```
so it should have gone for the default environment.  It seems to have done so in the other 4 bash scripts but not this one.

xylar commented:
@forsyth2, let me know when you'd like me to test again.  We're almost there...

forsyth2 commented:
@xylar I think it should be fine for you to test again.

@tomvothecoder and I are still trying to figure out why I keep getting `ModuleNotFoundError: No module named 'configobj'`.  Your `conda config` suggestions don't seem to work.

xylar commented:
@forsyth2, the next possibility is something in your `.condarc`, `.conda` or `.local` or similar.  Didn't you run into trouble like this awhile back?

tomvothecoder commented:
> @xylar I think it should be fine for you to test again.
> 
> @tomvothecoder and I are still trying to figure out why I keep getting `ModuleNotFoundError: No module named 'configobj'`. Your `conda config` suggestions don't seem to work.

Already sent this to @forsyth2, but I got it working using your steps @xylar. I had to make sure I was building `zppy` using the `base` environment, and not a custom environment. 

```
# Make sure you are in the base environment, not a custom one!
conda activate

# Update conda
conda update conda

# Install conda-build if you haven't already
conda install conda-build

# Build the package locally
conda build conda

# Create the environment with the local package
conda create -y -n zppy_package --use-local zppy

# Activate environment
conda activate zppy_package

# List packages, should have sub-dependencies that were installed through zppy
conda list
```

xylar commented:
@tomvothecoder, yes, you always have to build in the base environment I think.  Good call.

xylar commented:
@forsyth2, my test worked great this time.  Let me know if you need any other testing.  I'll obviously test out a release (or a release candidate) in E3SM-Unified if you want.

forsyth2 commented:
@xylar Awesome, thanks! So, is this pull request good to merge [once I 1) get my conda environment working and 2) squash the commits]?

xylar commented:
As far as I'm concerned, it's ready to merge.

forsyth2 commented on conda/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+  version: {{ version }}
+
+source:
+  #url: https://github.com/E3SM-Project/{{ name }}/archive/v{{ version }}.tar.gz
+  #sha256: ab6dee537f8252a9f244f01b725046dd64748d68
+  path: ..
```
```
  url: https://github.com/E3SM-Project/{{ name }}/archive/v{{ version }}.tar.gz
  sha256: ab6dee537f8252a9f244f01b725046dd64748d68
  #path: ..
```

forsyth2 commented on zppy/mpas_analysis.py:
```diff
@@ -1,7 +1,7 @@
 import os
-
-from utils import getTasks, getYears, submitScript, checkStatus
+from importlib import resources
```
Remove

forsyth2 reviewed (commented):
I reviewed the code again. I'm currently running `zppy -c` on a more complete configuration file. If that finishes successfully, I will squash the commits and merge.

For reference, I was able to get my `conda` environment working by doing the following:
```
rm -rf ~/miniconda3/
cd # cd into home directory
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash ./Miniconda3-latest-Linux-x86_64.sh
conda env list # only base environment is listed
cd <zppy-directory>
conda config --add channels conda-forge
conda config --set channel_priority strict
conda activate
conda update conda
conda install conda-build
conda build conda
conda create -y -n zppy_package --use-local zppy
conda activate zppy_package
conda list # configobj is now listed
cd ..
zppy -c 20210122.v2_test01.piControl.ne30pg2_EC30to60E2r2-1900_ICG.chrysalis.zppy_executable_simple.cfg
```
